### PR TITLE
Add device programmer subsystem stub code

### PR DESCRIPTION
**Pull request by guanyung-rs**
> ### Motivate of the pull request
> - [ ] To address an existing issue. If so, please provide a link to the issue: <issue id>
> - [x] Breaking new feature. If so, please describe the details in the description part.

> ### Describe the technical details
Created stub code for device programming subsystem
- Added DeviceProgrammer.cpp and header file to support following tcl commands
  - program, load_bitstream_file, verify
- Added ProgramDevice method for customization of the device programming flow

> #### What is currently done? (Provide issue link if applicable)
> 
> 
> 
> 
> 
>
> #### What does this pull request change?
> 
> 
> 
> 
> 

> ### Which part of the code base require a change
> 
> - [ ] Library: <Specify the library name>
> - [ ] Plug-in: <Specify the plugin name>
> - [x] Engine
> - [ ] Documentation
> - [ ] Regression tests
> - [ ] Continous Integration (CI) scripts

> ### Impact of the pull request

> - [ ] Require a change on Quality of Results (QoR)
> - [ ] Break back-compatibility. If so, please list who may be influenced.

### CODE REVIEW

**Comment by ravic-rs** [comment]
1. Needs a dummy batch tcl test/script with these functions.
2. Test plan for all repositories including verification of device programming in CI (likely in upstream repositories).

**Comment by guanyung-rs** [comment]
> 1. Needs a dummy batch tcl test/script with these functions.
> 2. Test plan for all repositories including verification of device programming in CI (likely in upstream repositories).

For #1, I added `program` tcl command into tests/TestGui/compiler_flow.tcl as part of testing

File: src/Compiler/CompilerDefines.h [hunk 1]
```diff
@@ -111,6 +111,7 @@ static constexpr const char *TIMING_ANALYSIS_LOG{"timing_analysis.rpt"};
 static constexpr const char *TA_TIMING_LOG{"post_ta_timing.rpt"};
 static constexpr const char *POWER_ANALYSIS_LOG{"power_analysis.rpt"};
 static constexpr const char *BITSTREAM_LOG{"bitstream.rpt"};
+static constexpr const char *PROGRAMDEVICE_LOG{"PROGRAMDEVICE.rpt"};
```
Comment (ravic-rs) [inline src/Compiler/CompilerDefines.h:114]: Remove this line, it's only for piping backend generated logs/std to a log file we display in GUI.

File: tests/TestGui/compiler_flow.tcl [hunk 2]
```diff
@@ -28,4 +28,5 @@ route
 sta
 power
 bitstream
+program
```
Comment (ravic-rs) [inline tests/TestGui/compiler_flow.tcl:31]: program_device

**Review by ravic-rs — CHANGES_REQUESTED** [review]
Some conflicts needs to be resolved due to recent merges in same files.

**Comment by alain-rs** [comment]
@chungshien please resolve the conflicts so we can merge this PR.

File: src/Compiler/Compiler.cpp [hunk 3]
```diff
@@ -427,8 +427,9 @@ bool Compiler::RegisterCommands(TclInterpreter* interp, bool batchMode) {
   }
   m_simulator->RegisterCommands(m_interp);
   m_IPGenerator->RegisterCommands(interp, batchMode);
-  // m_DesignQuery->RegisterCommands(interp, batchMode);
+  m_DesignQuery->RegisterCommands(interp, batchMode);
   m_deviceProgrammer->RegisterCommands(interp, batchMode);
+  m_DesignQuery->RegisterCommands(interp, batchMode);
```
Comment (guanyung-rs) [inline src/Compiler/Compiler.cpp]: This is redundant. Line 430 and 432 is the doing the same thing.

**Comment by chungshien** [comment]
@alain-rs - done. Had helped the team resolve the merge conflicts

@guanyung-rs - please double check when you are online and have time

But to be done, I am not sure I like the way the implementation is done. 
Currently it just "mimic" how designquery feature (GUI centric) is done and replicate the same for DeviceProgrammer. 

However DeviceProgrammer is not GUI feature, it is backend engine + command line executable support. The implementation include unnecessary dependencies/libraries that DeviceProgrammer does not need.

To me, this is not a clean implementation - a clean coding should only include the needed dependencies (header, library, etc).

**Comment by chungshien** [comment]
The windows fail with "D:\a\FOEDAG\FOEDAG\src\DesignRuns\..\..\src\DesignQuery/DesignQuery.h(31): fatal error C1083: Cannot open include file: 'nlohmann_json/json.hpp': No such file or directory"

But I do not change (add or delete or modify) anything that related to designquerry.  

@guanyung-rs  can you double check if the current changelist is exact your original implement after I fix the conflict?

**Comment by guanyung-rs** [comment]
> The windows fail with "D:\a\FOEDAG\FOEDAG\src\DesignRuns....\src\DesignQuery/DesignQuery.h(31): fatal error C1083: Cannot open include file: 'nlohmann_json/json.hpp': No such file or directory"
> 
> But I do not change (add or delete or modify) anything that related to designquerry.
> 
> @guanyung-rs can you double check if the current changelist is exact your original implement after I fix the conflict?

I am looking into the build issue in MSVC.

**Comment by guanyung-rs** [comment]
@ravic-rs, FYA, the CI build checks have passed. Please review.

**Comment by alain-rs** [comment]
The GUI needs a proxy (This subsystem) for the actual job done by a separate executable "Device programmer".
On the GUI side, this initial PR seems OK.

**Pull request merged**